### PR TITLE
swap cheque exchange protocol and settlement

### LINKED ISSUE #656: Swap settlement and chequebook integration
# Task

Implement swap chequebook as a settlement layer.

* [x] Introduce a swap settlement protocol to exchange cheques
* [x] Introduce a swap `Settlement` implementation
  * [x] signs and stores cheques and sends them when payment is required
  * [x] notifies accounting after receiving and verifying cheque
  * [x] verify counterparty chequebook on first encounter
  * [x] store received cheques so they can be cashed later

### PR DESCRIPTION
**Pull request by ralph-pichler**
adds the swap cheque exchange protocol and settlement layer.

additions:
* a new component `ChequeStore` that stores and verifies cheques received from foreign chequebooks
* `Chequebook` can now return last cheque for a beneficiary as well as all cheques
* (Swap) `Addressbook` which maps between peer address, beneficiary address and chequebook address
* `swap protocol` consisting of a message for cheques and a handshake to exchange beneficiary information
* `Swap` connecting everything together

unit tests still missing for api getter functions and the swarm protocol itself. Note that there is a missing liquidity check on the cheque issuing side (the node will happily sign and send bouncing cheques). As this PR is already huge and this logic is a bit complex this should be done in a separate PR.

closes #656

### CODE REVIEW

File: pkg/settlement/swap/chequebook/chequestore.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
+package chequebook
```
Comment (janos) [inline pkg/settlement/swap/chequebook/chequestore.go:1]: Add copyright header.

File: pkg/settlement/swap/chequebook/chequestore_test.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
+package chequebook_test
```
Comment (janos) [inline pkg/settlement/swap/chequebook/chequestore_test.go:1]: Add copyright header.

File: pkg/settlement/swap/chequebook/chequestore_test.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+		Cheque: chequebook.Cheque{
+			Beneficiary:      beneficiary,
+			CumulativePayout: cumulativePayout,
+			Chequebook:       chequebookAddress,
+		},
+		Signature: sig,
+	})
+	if err == nil {
+		t.Fatal("accepted cheque from invalid chequebook")
+	}
```
Comment (janos) [inline pkg/settlement/swap/chequebook/chequestore_test.go:262]: This check is not needed, it is covered with the next if statement.

File: pkg/settlement/swap/chequebook/chequestore_test.go [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+		Cheque: chequebook.Cheque{
+			Beneficiary:      beneficiary,
+			CumulativePayout: cumulativePayout,
+			Chequebook:       chequebookAddress,
+		},
+		Signature: sig,
+	})
+	if err == nil {
+		t.Fatal("accepted cheque with invalid signature")
+	}
```
Comment (janos) [inline pkg/settlement/swap/chequebook/chequestore_test.go:311]: This check is not needed, it is covered with the next if statement.

File: pkg/settlement/swap/chequebook/chequestore_test.go [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+		Cheque: chequebook.Cheque{
+			Beneficiary:      beneficiary,
+			CumulativePayout: cumulativePayout,
+			Chequebook:       chequebookAddress,
+		},
+		Signature: sig,
+	})
+	if err == nil {
+		t.Fatal("accepted cheque from illiquid chequebook")
+	}
```
Comment (janos) [inline pkg/settlement/swap/chequebook/chequestore_test.go:360]: This check is not needed, it is covered with the next if statement.

File: pkg/settlement/swap/metrics.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	// to be able to return them by Metrics()
+	// using reflection
+	TotalReceived prometheus.Counter
+	TotalSent     prometheus.Counter
+}
+
+func newMetrics() metrics {
+	subsystem := "swap"
+
+	return metrics{TotalReceived: prometheus.NewCounter(prometheus.CounterOpts{
```
Comment (janos) [inline pkg/settlement/swap/metrics.go:23]: New line for TotalReceived.

File: pkg/settlement/swap/swap.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+// ReceiveCheque is called by the swap protocol if a cheque is received.
+func (s *Service) ReceiveCheque(ctx context.Context, peer swarm.Address, cheque *chequebook.SignedCheque) (err error) {
+	// check this is the same chequebook for this peer as previously
+	expectedChequebook, known, err := s.addressbook.Chequebook(peer)
+	if err != nil {
+		return err
+	}
+	if known {
+		if expectedChequebook != cheque.Chequebook {
```
Comment (janos) [inline pkg/settlement/swap/swap.go:62]: if statements can be combined.

File: pkg/settlement/swap/swap_test.go [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+		chequebookService,
+		mockchequestore.NewChequeStore(),
+		addressbook,
+		networkID,
+	)
+
+	err := swap.Pay(context.Background(), peer, amount)
+	if err == nil {
+		t.Fatal("expected error")
+	}
```
Comment (janos) [inline pkg/settlement/swap/swap_test.go:365]: This check is covered with the next one.

File: pkg/settlement/swap/swap_test.go [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+
+func TestPayIssueError(t *testing.T) {
+	logger := logging.New(ioutil.Discard, 0)
+	store := mockstore.NewStateStore()
+
+	amount := uint64(50)
+	beneficiary := common.HexToAddress("0xcd")
+
+	peer := swarm.MustParseHexAddress("abcd")
+	reject := errors.New("reject")
```
Comment (janos) [inline pkg/settlement/swap/swap_test.go:335]: reject -> errReject

File: pkg/settlement/swap/swap_test.go [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+		addressbook,
+		networkID,
+	)
+
+	err := swap.Pay(context.Background(), peer, amount)
+	if err == nil {
+		t.Fatal("expected error")
+	}
+	if !errors.Is(err, reject) {
+		t.Fatal("wrong error")
```
Comment (janos) [inline pkg/settlement/swap/swap_test.go:367]: Log what is the wrong error.

File: pkg/settlement/swap/swap_test.go [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+		networkID,
+	)
+
+	err := swapService.Pay(context.Background(), peer, amount)
+	if err == nil {
+		t.Fatal("expected error")
+	}
+	if !errors.Is(err, swap.ErrUnknownBeneficary) {
+		t.Fatal("wrong error")
+	}
```
Comment (janos) [inline pkg/settlement/swap/swap_test.go:403]: Same as above.

File: pkg/settlement/swap/swap_test.go [hunk 12]
```diff
@@ -0,0 +1,505 @@
...
+		mockchequebook.NewChequebook(),
+		mockchequestore.NewChequeStore(),
+		&addressbookMock{},
+		networkID,
+	)
+
+	err := swapService.Handshake(peer, beneficiary)
+	if err == nil {
+		t.Fatal("accepted invalid beneficiary")
+	}
```
Comment (janos) [inline pkg/settlement/swap/swap_test.go:501]: Not needed check.

**Review by janos — COMMENTED** [review]
This is a very large PR that is not ideal for a good review, especially close to the release. Even the code quality is very good, I cannot say for sure that I managed to give it a very good review. The design is very nice, the only thing that I would like to try is to remove circular dependencies between settlements and accounting and swap and swap protocol. We can try to work on that later, if there is time.

File: pkg/settlement/swap/addressbook.go [hunk 13]
```diff
@@ -0,0 +1,129 @@
+package swap
```
[the remark this answers is not in the document]
Comment (ralph-pichler) [reply pkg/settlement/swap/addressbook.go:5]: fixed

File: pkg/settlement/swap/chequebook/chequestore.go (hunk 1 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/chequebook/chequestore.go:5]: fixed

File: pkg/settlement/swap/chequebook/chequestore_test.go (hunk 2 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/chequebook/chequestore_test.go:5]: fixed

File: pkg/settlement/swap/chequebook/chequestore_test.go (hunk 3 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/chequebook/chequestore_test.go]: fixed

File: pkg/settlement/swap/chequebook/chequestore_test.go (hunk 4 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/chequebook/chequestore_test.go]: fixed

File: pkg/settlement/swap/chequebook/chequestore_test.go (hunk 5 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/chequebook/chequestore_test.go]: fixed

File: pkg/settlement/swap/metrics.go (hunk 6 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/metrics.go]: fixed

File: pkg/settlement/swap/swap.go (hunk 7 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/swap.go]: fixed

File: pkg/settlement/swap/swap_test.go (hunk 8 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/swap_test.go]: fixed

File: pkg/settlement/swap/swap_test.go (hunk 9 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/swap_test.go]: fixed

File: pkg/settlement/swap/swap_test.go (hunk 10 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/swap_test.go]: fixed

File: pkg/settlement/swap/swap_test.go (hunk 11 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/swap_test.go]: fixed

File: pkg/settlement/swap/swap_test.go (hunk 12 above)
Comment (ralph-pichler) [reply pkg/settlement/swap/swap_test.go]: fixed

**Review by janos — APPROVED** [review]
Thanks @ralph-pichler for a very well written implementation.

**Review by pradovic — APPROVED** [review]
Very nice, @ralph-pichler! A bit big PR, so it's a bit hard to review, but excellent work!

**Pull request merged**